File: core/src/main/java/com/dmi/perfectreader/book/format/PageConfig.java
package com.dmi.perfectreader.book.format;

import javax.annotation.concurrent.Immutable;

import static com.google.common.base.Preconditions.checkArgument;

@Immutable
public class PageConfig {
    private final float width;
    private final float height;
    private final float paddingTop;
    private final float paddingRight;
    private final float paddingBottom;
    private final float paddingLeft;

    public PageConfig(float width, float height, float paddingTop, float paddingRight, float paddingBottom, float paddingLeft) {
        checkArgument(width >= 0);
        checkArgument(height >= 0);
        checkArgument(width + paddingLeft + paddingRight >= 0);
        checkArgument(height + paddingTop + paddingRight >= 0);
        this.width = width;
        this.height = height;
        this.paddingTop = paddingTop;
        this.paddingRight = paddingRight;
        this.paddingBottom = paddingBottom;
        this.paddingLeft = paddingLeft;
    }

    public float width() {
        return width;
    }

    public float height() {
        return height;
    }

    public float paddingTop() {
        return paddingTop;
    }

    public float paddingRight() {
        return paddingRight;
    }

    public float paddingBottom() {
        return paddingBottom;
    }

    public float paddingLeft() {
        return paddingLeft;
    }
}

File: app/src/main/kotlin/com/dmi/perfectreader/book/content/common/PageConfig.kt
package com.dmi.perfectreader.book.content.common

import com.dmi.perfectreader.MainContext
import com.dmi.perfectreader.book.pagination.page.Page
import com.dmi.perfectreader.settings.Settings
import com.dmi.util.graphic.SizeF
import com.dmi.util.graphic.shrink
import org.jetbrains.anko.displayMetrics

class PageConfig(
        context: MainContext,
        val size: SizeF,
        settings: Settings = context.settings,
        density: Float = context.android.displayMetrics.density,
        val paddings: Page.Paddings = Page.Paddings(
                settings.format.paddingLeftDip,
                settings.format.paddingRightDip,
                settings.format.paddingTopDip,
                settings.format.paddingBottomDip
        ) * density,
        val footer: Footer? = if (settings.screen.footerEnabled) Footer(context) else null,
        footerExtraSpace: Float = if (footer != null) (footer.height + footer.paddingBottom) else 0F,
        val contentSize: SizeF = size.shrink(paddings.left + paddings.right, paddings.top + paddings.bottom + footerExtraSpace),
        val textGammaCorrection: Float = settings.theme.textGammaCorrection
) {
    class Footer(
            context: MainContext,
            settings: Settings = context.settings,
            density: Float = context.android.displayMetrics.density,
            val height: Float = settings.screen.footerHeightEm * settings.font.sizeDip * density,
            val itemDistance: Float = 20 * density,
            val paddingBottom: Float = settings.screen.footerPaddingBottomPercent * settings.format.paddingBottomDip * density,
            val pageNumber: Boolean = settings.screen.footerPageNumber,
            val numberOfPages: Boolean = settings.screen.footerNumberOfPages,
            val chapter: Boolean = settings.screen.footerChapter
    )
}